### ISSUE TITLE
The framework default Gunicorn configs do not make sense for Gen2 environments

### ISSUE THREAD

**Issue by xSAVIKx** [GoogleCloudPlatform/functions-framework-python#241]
Hey, so we're extensively using the framework to create our GCF and Cloud Run services and it is very easy to use which is great.

But, when it comes to fine-tuning your performance things go rogue. E.g. we have a need for ~5 GB RAM in Cloud Run for some services, so effectively are 2 vCPUs already while the hard-coded gunicorn configs force only a single worker https://github.com/GoogleCloudPlatform/functions-framework-python/blob/586cc4e0833e2e9c82e685c8118be8c3187bd990/src/functions_framework/_http/gunicorn.py#L20-L27

And this is gonna be the same issue with GCF gen2 where you can use bigger instances in terms of memory but they also come with more vCPU for which you are paying but not using.

I suggest we either keep the default and let people configure/adjust Gunicorn params or implement a smart solution for picking appropriate defaults (at least based on the number of vCPU available).

**Comment by xSAVIKx** [comment]
Any updates?

**Comment by HKWinterhalter** [comment]
Hi @xSAVIKx, thank you for the suggestion. If I understand correctly, your concern is that there will be only a single worker, which would be a bottleneck and wasteful for higher vCPU counts.

I think the missing piece here is that the memory and vCPU settings actually apply to a Function deployment, which itself can be many instances each with a worker and many threads. The number of instances scale up and down based on need with lower and upper bounds being something you can configure. See: https://cloud.google.com/functions/docs/configuring/min-instances. 
I believe you could try setting max and min instances to effectively configure a Gunicorn worker count range.

The reason the thread setting is 1024 is because the concurrency feature, if turned on, allows many requests to be served by a single instance. See: https://cloud.google.com/functions/docs/configuring/concurrency

Perhaps you could try playing around with the about controls for performance tuning.

As a reminder, Cloud Functions is meant to be a more managed offering with some controls unavailable and vCPU as purposefully an abstraction. If you still need more control, you may find it with Cloud Run - for example, its docs have a section about optimizing Python applications and Gunicorn: https://cloud.google.com/run/docs/tips/python#optimize_gunicorn

**Comment by xSAVIKx** [comment]
Hi @HKWinterhalter. Thanks for the explanations, but please clarify a couple of things.

A Function deployment still has vCPU/RAM resources available to each Functions server (instance), right? Are you saying that if I pick a gen2 GCF with 8 vCPU, it will use all 8 vCPUs with the current Functions Framework setup?

Here's an example deployment screen from the GCP console:
![image](https://github.com/GoogleCloudPlatform/functions-framework-python/assets/5675689/a2573052-c532-4c72-9d4c-86e5736e0c46)

**Comment by HKWinterhalter** [comment]
After further investigation, your initial understanding was correct. I am marking this and your suggestions as an enhancement request. 

(I have also editted my original response as to not confuse future readers).

Thank you!

**Comment by looztra** [comment]
> [@xSAVIKx](https://github.com/xSAVIKx) I believe you can configure the Gunicorn params through env variables now

definitely, cf https://github.com/GoogleCloudPlatform/functions-framework-python/blob/main/src/functions_framework/_http/gunicorn.py#L39

```python
class GunicornApplication(gunicorn.app.base.BaseApplication):
    def __init__(self, app, host, port, debug, **options):
        threads = int(os.environ.get("THREADS", (os.cpu_count() or 1) * 4))

        global TIMEOUT_SECONDS
        TIMEOUT_SECONDS = int(os.environ.get("CLOUD_RUN_TIMEOUT_SECONDS", 0))

        self.options = {
            "bind": "%s:%s" % (host, port),
            "workers": int(os.environ.get("WORKERS", 1)),
            "threads": threads,
            "loglevel": os.environ.get("GUNICORN_LOG_LEVEL", "error"),
            "limit_request_line": 0,
        }
 ```
